tensorboard/webapp/widgets/custom_modal/custom_modal.ts

```diff
@@ -0,0 +1,180 @@
+import {
+  ApplicationRef,
+  Injectable,
+  TemplateRef,
+  ViewContainerRef,
+} from '@angular/core';
+import {CustomModalComponent} from './custom_modal_component';
+
+/**
+ * Enables dynamic creation of modal components.
+ *
+ * # Prerequisites
+ * App root component must define a ViewContainerRef named `modalViewContainerRef`
+ * e.g.:
+ *
+ * ```
+ * // Template file
+ * <div #modal_container></div>
+ * ```
+ *
+ * ```
+ * // Root Component definition
+ * class MyAppRoot {
+ *   @ViewChild('modal_container', {read: ViewContainerRef})
+ *   readonly modalViewContainerRef!: ViewContainerRef;
+ *   ...
+ * }
+ * ```
+ *
+ * # Usage
+ * Define a modal using an ng-template:
+ * ```
+ * <ng-template #myModalTemplate>
+ *   <custom-modal (onOpen)="doSomething" (onClose)="doSomethingElse">
+ *     <my-awesome-modal-content
+ *      [input1]="input1"
+ *      (output1)="output1"
+ *     >
+ *     </my-awesome-modal-content>
+ *   </custom-modal>
+ * </ng-template>
+ * ```
+ *
+ * Define a ViewChild to reference the template in the component file:
+ * ```
+ * // my_component.ts
+ * ...
+ * @ViewChild('myModalTemplate', {read: TemplateRef})
+ * myModalTemplate!: TemplateRef<unknown>;
+ * ...
+ * ```
+ *
+ * Inject CustomModal into the component
+ * ```
+ * // my_component.ts
+ * ...
+ * constructor(private readonly customModal: CustomModal) {}
+ * ...
+ * ```
+ *
+ * To create a modal, call createAtPosition():
+ * ```
+ * // my_component.ts
+ * ...
+ * onSomeButtonClick() {
+ *   this.customModal.createAtPosition(this.myModalTemplate, {x: 100, y: 100});
+ * }
+ * ...
+ * ```
+ *
+ * ## Important note
+ * runChangeDetection() must be called after view checked to prevent annoying
+ * https://angular.io/errors/NG0100 errors when using input bindings: this
+ * will run embedded view change detection in the correct order. Note that
+ * omitting this will not affect actual modal behavior.
+ * ```
+ * // my_component.ts
+ * ngAfterViewChecked() {
+ *   this.customModal.runChangeDetection();
+ * }
+ * ```
+ *
+ * # Testing
+ * A similar setup is required for testing modal behavior from another component.
+ *
+ * Define a separate root component to hold the modal container and ref:
+ * ```
+ * @Component({
+ *   ...,
+ *   template: `
+ *   <actual-component-to-test></actual-component-to-test>
+ *   <div #modal_container></div>
+ *   `
+ * })
+ * class TestableComponent {
+ *   @ViewChild('modal_container', {read: ViewContainerRef})
+ *   readonly modalViewContainerRef!: ViewContainerRef;
+ *
+ *   constructor(readonly customModal: CustomModal) {}
+ *   ...
+ * }
+ * ```
+ *
+ * Before each test (e.g. in beforeEach), make sure to add the TestableComponent
+ * to the app component list:
+ * ```
+ * ...
+ * const fixture = TestBed.createComponent(TestableComponent);
+ * const appRef = TestBed.inject(ApplicationRef);
+ * appRef.components.push(fixture.componentRef);
+ * ```
+ */
+@Injectable({providedIn: 'root'})
+export class CustomModal {
```

Comment:
Looks good. The problems I was having with the column selector in data table seem to have been solved!

I was wondering, though, could we have instead used Angular Material Overlay? https://material.angular.io/cdk/overlay/overview
Is CustomModal trying to solve the same problem as Overlay?

Reply:
Oh wow, didn't know about this! Yes I'd say the Overlay problem scope is basically equivalent to CustomModal's.

I immediately tried this approach out, as it's clearly more preferable to use a mature, featureful library if possible. Unfortunately, the dynamic nature of our clickable affordances doesn't seem to be compatible with how Overlay works.

To elaborate: Overlay usage ([ex](https://material.angular.io/cdk/overlay/examples)) seems to require referring to the clickable affordance by template variable (e.g. `#trigger`), but there is no straightforward way we can implement this for our *ngFor based column headers. Also, it seems that dynamically setting template variables won't be possible for the foreseeable future: https://github.com/angular/angular/issues/33233

Reply:
Overlay also contains a programmatic interface. Does that make it usable?

Here I migrated the filterbar_component to use overlay:
https://github.com/bmd3k/tensorboard/commit/cc7ebdd933744173d3969748b142c004b354ad86

I recognize that filterbar is the easiest of the usages of CustomModal but hopefully this points you in a useful direction. I admit that Overlay might still not be the right fit but I think it's worth exploring further.

Some other things to consider:

* custom_modal.ts could still exist as a service that wraps Overlay with some stuff specific to context menus. Maybe it's actually custom_menu.ts or context_menu.ts?

* once you've created an overlay for the top level of a menu, is it possible to just reuse that overlay for submenus rather than creating a new overlay for each descendent in the hierarchy?

Reply:
The code has been updated to use Overlay! Note that Overlay features make the CustomModalComponent redundant; we can solely use the CustomModal service now.

Some advantages of Overlay include:
- Modals can now follow trigger elements when the page is resized
- Less boilerplate (no extra `<custom-modal>` required in templates; no extra change detection management needed)
- Easy opportunities for future customization (e.g. configure modal behavior on scroll)

I've added some convenience logic in CustomModal to make Overlays work well with our nested modals requirements, along with some new tests. PTAL! +cc @rileyajones 

P.S.
> is it possible to just reuse that overlay for submenus rather than creating a new overlay for each descendent in the hierarchy?

I've found that each overlay must specify its own separate position, which means that each uniquely positioned modal needs its own overlay.

Reply:
Excellent. Thanks for bearing with me. I'm glad it actually went somewhere.